quil-rs/src/instruction/calibration.rs

```diff
@@ -9,6 +9,12 @@ use crate::{
 
 use super::write_qubit_parameters;
 
+pub trait CalibrationSignature {
+    type Signature: PartialEq;
+
+    fn signature(&self) -> Self::Signature;
+}
+
```

Comment:
If you do something like this, we could return references and wouldn't need to clone in the implementations (not tested):

```suggestion
pub trait CalibrationSignature {
    type Signature<'a>: PartialEq;

    fn signature<'a>(&'a self) -> Self::Signature<'a>;
}
```

Reply:
Nice, I recall trying this without any luck. Will give it another shot.

Reply:
Okay, I found the issue again. The provided snippet isn't valid, as Rust doesn't recognize the lifetime `'a` without defining it as a generic parameter to the trait. When you do provide it, it complains that the `'a` in `Signature` shadows the parameter.

I can get further with this definition:

```rs
pub trait CalibrationSignature<'a> {
    type Signature: PartialEq + 'a;

    fn signature(&'a self) -> Self::Signature;
}
```

But I can't figure out how to convince Rust that the lifetimes in the replace method is okay:

```rs
impl<'a, T> CalibrationSet<T>
where
    T: CalibrationSignature<'a>,
{
    // ...
    pub fn replace(&mut self, value: T) -> Option<T> {
        if let Some(index) = self.signature_position(value.signature()) {
            let replaced = std::mem::replace(&mut self.data[index], value);
            Some(replaced)
        } else {
            self.data.push(value);
            None
        }
    }
}
```

Errors with:

```sh
error[E0597]: `value` does not live long enough
   --> quil-rs/src/program/calibration_set.rs:109:54
    |
58  | impl<'a, T> CalibrationSet<T>
    |      -- lifetime `'a` defined here
...
105 |     pub fn replace<'b>(&'b mut self, value: T) -> Option<T>
    |                                      ----- binding `value` declared here
...
109 |         if let Some(index) = self.signature_position(value.signature()) {
    |                                                      ^^^^^------------
    |                                                      |
    |                                                      borrowed value does not live long enough
    |                                                      argument requires that `value` is borrowed for `'a`
...
116 |     }
    |     - `value` dropped here while still borrowed

error[E0505]: cannot move out of `value` because it is borrowed
   --> quil-rs/src/program/calibration_set.rs:110:69
    |
58  | impl<'a, T> CalibrationSet<T>
    |      -- lifetime `'a` defined here
...
105 |     pub fn replace<'b>(&'b mut self, value: T) -> Option<T>
    |                                      ----- binding `value` declared here
...
109 |         if let Some(index) = self.signature_position(value.signature()) {
    |                                                      -----------------
    |                                                      |
    |                                                      borrow of `value` occurs here
    |                                                      argument requires that `value` is borrowed for `'a`
110 |             let replaced = std::mem::replace(&mut self.data[index], value);
    |                                                                     ^^^^^ move out of `value` occurs here

error[E0505]: cannot move out of `value` because it is borrowed
   --> quil-rs/src/program/calibration_set.rs:113:28
    |
58  | impl<'a, T> CalibrationSet<T>
    |      -- lifetime `'a` defined here
...
105 |     pub fn replace<'b>(&'b mut self, value: T) -> Option<T>
    |                                      ----- binding `value` declared here
...
109 |         if let Some(index) = self.signature_position(value.signature()) {
    |                                                      -----------------
    |                                                      |
    |                                                      borrow of `value` occurs here
    |                                                      argument requires that `value` is borrowed for `'a`
...
113 |             self.data.push(value);
```

There are similar lifetime issues in signature_position(), and some of the trait implementations, but I suspect if we can fix this, we can fix the others.